### ISSUE TITLE
why define `[key: string]: object | any` on responses before the actual shape?

### ISSUE THREAD

**Issue by purge** [plaid/plaid-node#491]
Defining this completely makes having type checks completely pointless. (in 9.2.0)

```
let {data} = await plaidClient.institutionsGetById(...);
data.imadethisup // types as 'any'
```

**Comment by phoenixy1** [comment]
@purge I'm sorry to get to this a million years late. I'm not a node expert, but I believe the reason we do this is to make the client libraries robust to server changes -- we don't want people to have to update their client library any time we make e.g. an additive change to the server's responses. Definitely open to suggestions on alternative / better ways to do this.

**Comment by purge** [comment]
I think people would be fine upgrading their library when your api changes, thats why you're building it from your schema (and what every other api library does). injecting `any` breaks type safety.

**Comment by phoenixy1** [comment]
I definitely wish people would be fine with it! But my experience in practice is that quite a few of our API users are pretty reluctant to upgrade / update anything, including client libraries :-/. Given that our current API versioning system considers additive changes to be automatic/there's no way to opt out of them, it's tough to make it so that people's client libraries and the API are always perfectly in sync.

**Comment by ryanmsnyder** [comment]
I'm also curious why `[key: string]: object | any;` appears on seemingly every interface, which almost makes the entire plaid declaration file useless. I wouldn't think additive changes would cause any issues.

**Comment by ryanmsnyder** [comment]
I would love this fix! Not to beat a dead horse but `api.d.ts` has 537 occurrences of `[key: string]: object | any;`. That's a lot of type safety that's taken away.

**Comment by phoenixy1** [comment]
So after looking into it a bit more, I think the best we would be able to do is probably something like this, replacing the `any` with `unknown`:

https://github.com/bcherny/json-schema-to-typescript/pull/281

This doesn't solve the type safety issue 100%, and is also a bit of a trade-off, so I'm not 100% sure we'd want to adopt it (we still want to be friendly for users of older client library versions) but I'm running it past the team and we'll see what they think.

**Comment by donleistman** [comment]
+1, noticed this and would also appreciate stricter types here

**Comment by SimHoZebs** [comment]
[According to an answer to this StackOverflow question](https://stackoverflow.com/questions/76607165/why-does-interface-lose-its-properties-when-one-property-is-defined-with-index-s/76607495#76607495), `[key: string:] object|unknown` property destroys other properties from being inferred by TypeScript. In my case, tRPC is failing to infer the types properly.

As far as I understand, the reason for this general property is so developers can access properties that exist on the API, but are not defined on the client library (due to using an outdated client). I'm not sure if that is worth completely losing type-safety, considering developers should expect no change, including additive changes, if their library version has remained the same.

**Comment by purge** [comment]
```typescript
type result = {
  knownKey: string;
  knownKey2: number;
  [k: string]: unknown;
};
```

is generally how that is done.

**Comment by Ranhiru** [comment]
We came across the same issue and had to reexport `Strict` versions of the types for usage

See https://stackoverflow.com/a/68261113

```typescript
type Strict<T> = {
  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
};

export type AccountBaseStrict = Strict<AccountBase>;
export type AccountsGetResponseStrict = Strict<AccountsGetResponse>;
```

**Comment by zackdotcomputer** [comment]
I think for Plaid's team, I want to clarify the purpose of this adaptation. @phoenixy1, you said that it was a requirement was "being able to use outdated clients without the library generating errors when new properties are added" but that could mean two things? I agree with one way of reading that requirement: I should be able to keep compiling my code from v12 even if the deployed API is on v18. I think the thing that we (the posters in this thread) disagree with is the opposite - that developers should be able to write v18 code against the v12 library without THAT generating errors. If the only point of the library is to provide the functions that make the API calls but permit anything be done with the data returned, why bother having any typing at all?

By making the `interface` types this permissive, the team has overridden client developers' choices regarding typing. This is frustrating for two reasons:

Firstly, it makes it difficult to use your SDK to guide models and interfaces in our own code. I cannot now ensure my database representation of an account balance matches yours by saying `implements AccountBalance` because your `AccountBalance` interface declares that it could contain any string key in the future but my database strictly cannot. Instead, I now have to use a workaround like @Ranhiru to remove the permissiveness hack.

Secondly, it destroys the main purpose of typing - to catch errors in the otherwise formless space that is ES/JS. By adding this allowance, I must now move at a snail's pace when interacting with Plaid code to make sure your SDK is not allowing unnecessary mistakes to creep in. Is it `TransferAuthorizationPaymentRisk.warnings` or `TransferAuthorizationPaymentRisk.warning`? Both compile fine, but only one will ever work.

If, as you have suggested, there is a huge population of people who want to build v18 functionality against the v12 sdk for some reason (and they are unwilling to use the `as any` hacks in their own code), I am surprised to learn that. But if they do exist, then the way to support them would, in my opinion, be to roll the opposite of the `Strict` workaround above into the SDK itself: declare a utility type `Permissive<T> = T & { [key: string]: object | unknown }` and then use that where you use interfaces you want to be permissive instead of rolling it into the interface itself.

As an example, instead of making `AccountFiltersResponse` include this hack itself, have `LinkTokenGetMetadataResponse` contain a property `account_filters?: Permissive<AccountFiltersResponse>`. Then at least the `AccountFiltersResponse` interface exported from the SDK will be useful as-is rather than requiring us to hack usefulness back into. We still would run the risk that `LinkTokenGetMetadataResponse.account_filters.depsitory` would compile seamlessly (there's a typo in there), but at least we could mitigate that with a direct cast - `(LinkTokenGetMetadataResponse.account_filters as AccountFiltersResponse).depsitory` would flag the typo.

**Comment by phoenixy1** [comment]
@zackdotcomputer thanks so much for your comment! 

So, to back up and provide context on how we think about this:
The Plaid client libraries are created with the OpenAPI generator tool, which ingests an OpenAPI format file that specifies the API and outputs a client library.

The Plaid API is designed in such a way that it should be allowed to add new fields to objects without those being considered breaking changes. The way this property is represented in OpenAPI is via the `additionalProperties: true` attribute being applied to an object.

Back when we were originally writing these client libraries, we did not apply `additionalProperties: true` to objects. Indeed, the default behavior of OpenAPI is to assume `additionalProperties: true` if additionalProperties is not specified. However, none of the actual client library generators honored this part of the spec. So we had to add an explicit `additionalProperties: true` to represent how the API behaves. 

Most importantly, the Python client library broke in the following way if `additionalProperties: true` was not specified: if an API response object contained a field that was added in a later version of the API, the client library would throw an error because the field was unrecognized.

The current behavior of the Node client library reflects our OpenAPI generator's handling of `additionalProperties: true` for Node. (But obviously, we can change this by customizing the mustache files.)

The question of whether it's necessary to allow someone to actively use / send / interact with a field that's not explicitly supported by the existing client library is a subjective one. I would prefer to allow this, if possible -- upgrading the client library is a surprisingly large hurdle for a lot of our customers and the requirement to upgrade can be a barrier to getting them to adopt new features -- but it's not a strict requirement to allow that if the alternative is to cause a lot of pain for a lot of other customers in terms of type handling. The behavior of not-breaking existing integrations, of course, is non-negotiable. 

Anyway, I don't have enough knowledge about JavaScript types to actually evaluate your specific suggestion, and I think it's a thoughtful suggestion that deserves a thoughtful response, so I'm tagging in my colleague @jking-plaid, who has more expertise in JavaScript, to take a look and respond.

**Comment by zackdotcomputer** [comment]
Thank *you*, @phoenixy1 - that explanation makes it much clearer how we got here.

One other thing that occurred to me after my initial post was that I've seen this handled elsewhere using extensions of response types on the API functions themselves. So for example, currently the `bankTransferBalanceGet` API call is typed like this (where `BankTransferBalanceGetResponse` has the additional properties adaptation):

```
bankTransferBalanceGet(
  bankTransferBalanceGetRequest: BankTransferBalanceGetRequest, options?: any
): Promise<
  (axios?: AxiosInstance | undefined, basePath?: string | undefined) => AxiosPromise<BankTransferBalanceGetResponse>
>;
```

but I've seen other client libraries that would sign it like this:

```
bankTransferBalanceGet<ResponseType extends BankTransferBalanceGetResponse = BankTransferBalanceGetResponse>(
  bankTransferBalanceGetRequest: BankTransferBalanceGetRequest, options?: any
): Promise<
  (axios?: AxiosInstance | undefined, basePath?: string | undefined) => AxiosPromise<ResponseType>
>;
```

This would allow developers to explicitly add properties to the `BankTransferBalanceGetResponse` that they wanted to use but which were not included in their version of the SDK, by calling it as `bankTransferBalanceGet<BankTransferBalanceGetResponse & { additionalProperty: string }>(...)`. However, it leaves the default case strictly typed to the properties that existed in the response when that version of the SDK was exported.

Anyways, I am at this point trying to help us brainstorm a solution here. I hear your point that for some people in larger, more enterprise-grade codebases, they may have stricter rules around dependency upgrades that trap them on older versions of the SDK. Meanwhile my perspective, working at scrappier companies, is that we are able to upgrade to the latest releases but we have less complete QA processes than those teams, which means we're more dependent on tools like strict typing to ensure that we haven't made minor mistakes like typos in our property names. I'm hopeful we can find an adaptation that serves both needs.

**Comment by jking-plaid** [comment]
Hey all, after a bit of internal discussion, we've decided to remove index signatures from many of our types. This is available starting with the new `19.0.0` release. Please see [`CHANGELOG.md`](https://github.com/plaid/plaid-node/blob/master/CHANGELOG.md) for more details on upgrading. Hopefully this addresses the concerns here about stricter type safety!

**Comment by zackdotcomputer** [comment]
Thanks @jking-plaid and @phoenixy1 for engaging on this point. I just upgraded to v19 and confirmed that I was able to remove the `Strict` adaptation from my code thanks to it.
